app/src/main/java/ch/hikemate/app/model/elevation/ElevationServiceRepository.kt

```diff
@@ -47,67 +98,192 @@ class ElevationServiceRepository(private val client: OkHttpClient) : ElevationSe
+          if (coordinates.size < MAX_COORDINATES_PER_REQUEST) {
+            delay(CACHE_UPDATE_DELAY)
+            // If there were new requests in the meantime, we need to update the coordinates
+            mutex.withLock { coordinates = requests.flatMap { it.coordinates } }
+          }
+          mutex.withLock {
+            val chunks = coordinates.chunked(MAX_COORDINATES_PER_REQUEST)
+
+            for (chunk in chunks) {
+              // Create the JSON body for the request
+              val jsonString = buildJsonArray {
+                chunk.forEach { coordinate ->
+                  add(
+                      buildJsonArray {
+                        add(coordinate.lat)
+                        add(coordinate.lon)
+                      })
+                }
+              }
+
+              val request =
+                  okhttp3.Request.Builder().url(BASE_URL + jsonString.toString()).get().build()
+
+              // Callback for the network request, it also updates the cache
+              val onSuccessWithCache: (List<Double?>) -> Unit =
+                  parseCacheOnDataRetrieval(chunk, onSuccess)
+
+              val onFailureCallback: (Exception) -> Unit = { exception ->
+                CoroutineScope(dispatcher).launch {
+                  mutex.withLock { requests.forEach { request -> request.onFailure(exception) } }
+                  onFailure(exception)
+                }
+              }
+
+              client
+                  .newCall(request)
+                  .enqueue(ElevationServiceCallback(onSuccessWithCache, onFailureCallback))
             }
-          })
+          }
+        }
+  }
+
+  /**
+   * Parse the cache on data retrieval
+   *
+   * @param chunk The chunk of coordinates that were requested
+   * @param onSuccess The callback to be called when the request is successful
+   * @return A callback that will parse the cache on data retrieval
+   */
+  private fun parseCacheOnDataRetrieval(
+      chunk: List<LatLong>,
+      onSuccess: (List<Double>) -> Unit
+  ): (List<Double?>) -> Unit = { listOfElevation ->
+    val nonNullListOfElevation = listOfElevation.map { it ?: 0.0 }
+
+    CoroutineScope(dispatcher).launch {
+
+      // Update the cache with the new data
+      mutex.withLock {
+        chunk.forEachIndexed { index, it ->
+          cache[it] = ElevationCacheEntry(Date(), nonNullListOfElevation[index])
+        }
+
+        val processedRequests: MutableSet<ElevationRequest> = mutableSetOf()
+
+        // Call the onSuccess callback for each request
+        requests.forEach { request ->
+          val elevations = request.coordinates.map { cache[it]?.elevation }
+          val notNullElevations = elevations.filterNotNull()
+          if (notNullElevations.size != elevations.size) {
+            Log.e(
+                "ElevationServiceRepository",
+                "Some elevations are missing in cache, waiting for more data")
```

Comment:
Change: late comment, but this needs to be changed to `Log.i`, since it inevitably happens in the implementation and is not a problem at all.

Reply:
Yep! Good catch!